driver/serial.cpp

```diff
@@ -15,8 +15,10 @@ bool SERIAL::isBufferNull() {
 }
 
 void SERIAL::send(char a) {
-   while (!isBufferNull()){}
-   outb(port,a);
+//   while (!isBufferNull()){}
+   int i;
+   for(i = 0; i < 128 && !isBufferNull(); i++){}
```

Comment:
What happen  if the buffer is not null after i reach to 127?

Reply:
Khi buffer không null, nghĩa là có 1 thiết bị kết nối vào cổng serial , thì nó ngưng loop while. Nếu không có thiết bị khác kết nối đến cổng serial, thì nó sẻ while loop mãi. 
Vì thế mình thay bằng for loop,  sẻ không bị loop mãi dù cho có thiết bị khác kết nối hay không kết nối đến cổng serial .

Reply:
Mình thấy bạn chờ cho đến khi Buffer null mới send data. 
Nếu khi nếu thúc vòng for mà buffer chưa null, mình thấy bạn vẫn send data

Reply:
Đúng rồi.

Reply:
Vậy có phải sửa gì k bạn? Vì mình thấy vẫn send data khi chưa đảm bảo Buffer đã null

Reply:
Bạn có ý kiến hay hơn thì fix, ko để sau

Reply:
Mình nghĩ có thể dùng cơ chế re-try. Sẽ re-try trong khoảng vài lần, nếu buffer null thì send data và break vòng lặp. Nếu không null thì thôi

Reply:
Không biết có được ko. Vì trong hàm printSerial() có while hàm send().

Reply:
Hàm printSerial dùng while để print các ký tự trong msg nên mình nghĩ không ảnh hưởng

Reply:
re-try như nào bạn

Reply:
Mình nghĩ để vậy. 

void SERIAL::send(char a) {
   if (isBufferNull())
       outb(port,a);
}

Reply:
Chạy thấy vẫn lỗi

Reply:
hi @nguyenzung  .Mình đã test code này, thấy ổn.  

  while (!isBufferNull()){ KetNoi = 0;break;  }
  while (isBufferNull()){ KetNoi = 1;break;  }
  if (KetNoi == 0 && a == '\x20' )    {
              printf(" Ko co Ket Noi ");
                                                       } else if (KetNoi == 1)    {
                                                                                              outb(port,a);
                                                                                              }